src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs

```diff
@@ -497,19 +497,40 @@ public void AddMetadataReference(MetadataReference reference, AssemblyIdentity i
 
                 if (IsInGAC(peRef.FilePath) && identity != null)
                 {
+                    // Since the location of the reference is in GAC, need to use full identity name to find it again.
+                    // This typically happens when you base the reference off of a reflection assembly location.
                     _loadedProject.AddItem("Reference", identity.GetDisplayName(), metadata);
                 }
-                else
+                else if (IsReferenceAssembly(peRef.FilePath))
+                {
+                    // just use short name since this will be resolved by msbuild relative to the known reference assemblies.
+                    var fileName = identity != null ? identity.Name : Path.GetFileNameWithoutExtension(peRef.FilePath);
+                    _loadedProject.AddItem("Reference", fileName, metadata);
+                }
+                else // other location -- need hint to find correct assembly
                 {
                     string relativePath = FilePathUtilities.GetRelativePath(_loadedProject.DirectoryPath, peRef.FilePath);
-                    _loadedProject.AddItem("Reference", relativePath, metadata);
+                    var fileName = Path.GetFileNameWithoutExtension(peRef.FilePath);
+                    metadata.Add("HintPath", relativePath);
+                    _loadedProject.AddItem("Reference", fileName, metadata);
                 }
             }
         }
 
+        private static readonly string s_GAC = $"{Path.DirectorySeparatorChar}GAC_MSIL{Path.DirectorySeparatorChar}";
+        private static readonly string s_ALTGAC = $"{Path.AltDirectorySeparatorChar}GAC_MSIL{Path.AltDirectorySeparatorChar}";
+
         private bool IsInGAC(string filePath)
         {
-            return filePath.Contains(@"\GAC_MSIL\");
+            return filePath.Contains(s_GAC) || filePath.Contains(s_ALTGAC);
```

Comment:
Okay, I've update the implementation to use known gac & reference assembly locations